docs/design.md

```diff
@@ -0,0 +1,128 @@
+# Overview
+
+Concord-bft is an implementation of the [SBFT
+Protocol](https://arxiv.org/abs/1804.01626). SBFT implements State Machine
+Replication (SMR) in the byzantine model. It is a generic protocol, with
+pluggable interfaces for communication, storage, and state transfer. This
+document covers the implementation of concord-bft. When the implementation
+differs from the paper, it will be mentioned explicitly.
```

Comment:
All documents about code are to some degree aspirational, and they asymptotically approach correctness.  Documentation can always get out of date, and we should always have the goal of keeping it in sync.

However, saying that "we should" do something in the docs doesn't really makes sense IMO. We should just do it. If it gets out of date we should fix it. This is an implied contract with the reader. This comment does bring up an important question though: "What can we do to prevent the docs from getting out of sync?"

We probably should have a checklist for pull requests of things to look for when updating the code and docs. This would also be useful to help with determining reviewers. Ittai and I specifically discussed that we want core protocol changes to be run by researchers, and having a checklist would also help with this. Ideally, we could make the checklist part of the PR signoff process. The checklist must be completed before any PR can be merged. I think GitHub might be able to help us with this functionality. If not we can just enforce it via convention in code reviews, by having the author put it in the PR description.